test/framework/exec/docker.go

```diff
@@ -0,0 +1,360 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package exec
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/docker/go-connections/nat"
+)
+
+const (
+	httpProxy  = "HTTP_PROXY"
+	httpsProxy = "HTTPS_PROXY"
+	noProxy    = "NO_PROXY"
+)
+
+var dockerClient *client.Client
+
+// getDockerClient returns a new client connection for interacting with the Docker engine.
+func getDockerClient() (*client.Client, error) {
+	if dockerClient == nil {
+		var err error
+		dockerClient, err = client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+		if err != nil {
+			return nil, err
+		}
+	}
+	return dockerClient, nil
+}
+
+// SaveImage saves a Docker image to the file specified by dest.
+func SaveDockerImage(ctx context.Context, image, dest string) error {
+	docker, err := getDockerClient()
+	if err != nil {
+		return err
+	}
+
+	reader, err := docker.ImageSave(ctx, []string{image})
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+
+	tar, err := os.Create(dest)
+	if err != nil {
+		return err
+	}
+	defer tar.Close()
```

Comment:
We're probably doing this all across the code base. But I'm a bit concerned about just ignoring errors in those cases. I know it's not that critical in tests but let's discuss it :)

Can you take a look at: https://github.com/thanos-io/thanos/blob/main/docs/contributing/coding-style-guide.md#reliability

WDYT about the pattern they are using with `defer runutil.CloseWithErrCapture(&err, f, "close file")` ?

Reply:
Yeah, I think that would be great. Since that potentially hits several different areas outside of this change, I've filed https://github.com/kubernetes-sigs/cluster-api/issues/4586 to follow up on that.